R/bundleFiles.R

```diff
@@ -0,0 +1,261 @@
+    } else if (dirExists(target)) {
+      # a directory; explode it
+      contentPaths <- file.path(f, list.files(
+        target,
+        recursive = TRUE,
+        include.dirs = FALSE
+      ))
+
+      exploded <- c(exploded, contentPaths)
+      totalFiles <- totalFiles + length(contentPaths)
+      totalSize <- totalSize + sum(file_size(file.path(dir, contentPaths)))
+    } else {
+      # must be an ordinary file
+      exploded <- c(exploded, f)
+      totalFiles <- totalFiles + 1
+      totalSize <- totalSize + file_size(target)
+    }
+
+    enforceBundleLimits(dir, totalFiles, totalSize)
+  }
+
+  exploded
+}
+
+#' List Files to be Bundled
+#'
+#' @description
+#' Given a directory containing an application, returns the names of the files
+#' that by default will be bundled in the application. It works similarly to
+#' a recursive directory listing from [list.files()] but:
+#'
+#' * If the total size of the files exceeds the maximum bundle size,
+#'   `listBundleFiles()` will error. The maximum bundle size is controlled by
+#'    the `rsconnect.max.bundle.size` option.
+#' *  If the total size number of files exceeds the maximum number to be
+#'    bundled, `listBundleFiles()` will error. The maximum number of files in
+#'    the bundle is controlled by the `rsconnect.max.bundle.files` option.
+#' *  Certain files and folders that don't need to be bundled, such as
+#'    those containing internal version control and RStudio state, are
+#'    automatically excluded.
+#' *  You can exclude additional files by listing them in in a `.rscignore`
+#'    file. This file must have one file or directory per line (with path
+#'    relative to the current directory). It doesn't support wildcards, or
+#'    ignoring files in subdirectories.
+#'
+#' @param appDir Directory containing the application.
+#' @return Returns a list containing the following elements:
+#' * `totalFiles`: Total number of files.
+#' * `totalSize`: Total size of the files (in bytes).
+#' * `contents`: Paths to bundle, relative to `appDir`.
+#' @export
+listBundleFiles <- function(appDir) {
+  recursiveBundleFiles(appDir)
+}
+
+bundleFiles <- function(appDir) {
+  listBundleFiles(appDir)$contents
+}
+
+recursiveBundleFiles <- function(dir,
+                                 root_dir = dir,
+                                 depth = 0,
+                                 totalFiles = 0,
+                                 totalSize = 0) {
+  # generate a list of files at this level
+  contents <- list.files(dir, all.files = TRUE, no.. = TRUE, include.dirs = TRUE)
+  contents <- ignoreBundleFiles(dir, contents, depth = depth)
+
+  # Info for each file lets us know to recurse (directories) or aggregate (files).
+  is_dir <- dir.exists(file.path(dir, contents))
+  names(is_dir) <- contents
+
+  children <- character()
+  for (name in contents) {
+    if (isTRUE(is_dir[[name]])) {
+      out <- recursiveBundleFiles(
+        dir = file.path(dir, name),
+        root_dir = root_dir,
```

Comment:
nit: mixing argument name styles. camels hate snakes. i'm completely aware that i am actively throwing stones and simultaneously living in a glass house. 😜

Reply:
EmBrAcE ThE ChAoS